sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/conditionalExpressions.scala

```diff
@@ -340,3 +343,105 @@ object CaseKeyWhen {
     CaseWhen(cases, elseValue)
   }
 }
+
+/**
+ * A function that returns the index of expr in (expr1, expr2, ...) list or 0 if not found.
+ * It takes at least 2 parameters, and all parameters should be subtype of AtomicType or NullType.
+ * It's also acceptable to give parameters of different types. When the parameters have different
+ * types, comparing will be done based on type firstly. For example, ''999'' 's type is StringType,
+ * while 999's type is IntegerType, so that no further comparison need to be done since they have
+ * different types.
+ * If the search expression is NULL, the return value is 0 because NULL fails equality comparison
+ * with any value.
+ * To also point out, no implicit cast will be done in this expression.
+ */
+// scalastyle:off line.size.limit
+@ExpressionDescription(
+  usage = "_FUNC_(expr, expr1, expr2, ...) - Returns the index of expr in the expr1, expr2, ... or 0 if not found.",
+  extended = """
+    Examples:
+      > SELECT _FUNC_(10, 9, 3, 10, 4);
+       3
+      > SELECT _FUNC_('a', 'b', 'c', 'd', 'a');
+       4
+      > SELECT _FUNC_('999', 'a', 999, 9.99, '999');
+       4
+  """)
+// scalastyle:on line.size.limit
+case class Field(children: Seq[Expression]) extends Expression {
+
+  /** Even if expr is not found in (expr1, expr2, ...) list, the value will be 0, not null */
+  override def nullable: Boolean = false
+  override def foldable: Boolean = children.forall(_.foldable)
+
+  private lazy val ordering = TypeUtils.getInterpretedOrdering(children(0).dataType)
+
+  private val dataTypeMatchIndex: Array[Int] = children.zipWithIndex.tail.filter(
+    _._1.dataType.sameType(children.head.dataType)).map(_._2).toArray
+
+  override def checkInputDataTypes(): TypeCheckResult = {
```

Comment:
For the current implementation, I didn't do implicit cast to in accord with Hive.
For example:
`select field(3, '3', 3);` will return 2 in Hive, while it will return 1 if the implicit cast's done.
So I think maybe it's better not to do the implicit cast, and to allow that children have different types.